client/fingerprint_manager.go

```diff
@@ -73,43 +83,149 @@ func (fm *FingerprintManager) setupDrivers(drivers []string) error {
 			return err
 		}
 
-		detected, err := fm.fingerprint(name, d)
+		detected, err := fm.fingerprintDriver(name, d)
 		if err != nil {
 			fm.logger.Printf("[DEBUG] client.fingerprint_manager: fingerprinting for %v failed: %+v", name, err)
 			return err
 		}
 
+		if hc, ok := d.(fingerprint.HealthCheck); ok {
+			fm.runHealthCheck(name, hc)
+		} else {
+			// for drivers which are not of type health check, update them to have their
+			// health status match the status of whether they are detected or not.
+			healthInfo := &structs.DriverInfo{
+				Healthy:    detected,
+				UpdateTime: time.Now(),
+			}
+			if node := fm.updateNodeFromDriver(name, nil, healthInfo); node != nil {
+				fm.nodeLock.Lock()
+				fm.node = node
+				fm.nodeLock.Unlock()
+			}
+		}
+
+		go fm.watchDriver(d, name)
+
 		// log the fingerprinters which have been applied
 		if detected {
 			availDrivers = append(availDrivers, name)
 		}
-
-		p, period := d.Periodic()
-		if p {
-			go fm.run(d, period, name)
-		}
 	}
 
 	fm.logger.Printf("[DEBUG] client.fingerprint_manager: detected drivers %v", availDrivers)
 	return nil
 }
 
+// watchDrivers facilitates the different periods between fingerprint and
+// health checking a driver
+func (fm *FingerprintManager) watchDriver(d driver.Driver, name string) {
+	_, fingerprintPeriod := d.Periodic()
+	fm.logger.Printf("[DEBUG] client.fingerprint_manager: fingerprinting driver %s every %v", name, fingerprintPeriod)
+
+	var healthCheckPeriod time.Duration
+	hc, isHealthCheck := d.(fingerprint.HealthCheck)
+	if isHealthCheck {
+		req := &cstructs.HealthCheckIntervalRequest{}
+		var resp cstructs.HealthCheckIntervalResponse
+		hc.GetHealthCheckInterval(req, &resp)
+		if resp.Eligible {
+			fm.logger.Printf("[DEBUG] client.fingerprint_manager: health checking driver %s every %v", name, healthCheckPeriod)
+			healthCheckPeriod = resp.Period
+		}
+	}
+
+	t1 := time.NewTimer(fingerprintPeriod)
+	defer t1.Stop()
+	t2 := time.NewTimer(healthCheckPeriod)
+	defer t2.Stop()
+
+	for {
+		select {
+		case <-fm.shutdownCh:
+			return
+		case <-t1.C:
+			t1.Reset(fingerprintPeriod)
+			_, err := fm.fingerprintDriver(name, d)
+			if err != nil {
+				fm.logger.Printf("[DEBUG] client.fingerprint_manager: periodic fingerprinting for driver %v failed: %+v", name, err)
+			}
+		case <-t2.C:
+			if isHealthCheck {
```

Comment:
I feel like we shouldn't run the health check if detected isn't true. Otherwise you will log errors continuously, right?

I think you also skip the initial health check here and just rely on this loop: https://github.com/hashicorp/nomad/pull/3856/files#diff-8cf4b6aaa496508270af1cd14fdc1527R92

Reply:
Agree with this, but the second point would entail that all drivers that allow health checking will also be periodic. If that is always true, then we can do as you said and skip the initial check and only update on a periodic interval.